### ISSUE TITLE
solver.cpp : smoothed_loss becomes negative while loss stays positive

### ISSUE THREAD

**Issue by to3i** [BVLC/caffe#3029]
During training of a network with Euclidean loss layers using the adam solver the following happens in some cases when the initial loss is high:

```
I0904 14:04:10.916004 21891 solver.cpp:242] Iteration 0, loss = 2.38658e+18
I0904 14:04:10.916059 21891 solver.cpp:258]     Train net output #0: loss = 2.38647e+18 (* 1 = 2.38647e+18 loss)
I0904 14:04:10.916079 21891 solver.cpp:258]     Train net output #1: loss2 = 2.79839e+13 (* 4 = 1.11936e+14 loss)
I0904 14:04:10.916112 21891 solver.cpp:571] Iteration 0, lr = 0.0001
...
...
...
I0904 14:05:02.294297 21891 solver.cpp:242] Iteration 60, loss = 1.28282e+11
I0904 14:05:02.294333 21891 solver.cpp:258]     Train net output #0: loss = 1.68715e+11 (* 1 = 1.68715e+11 loss)
I0904 14:05:02.294359 21891 solver.cpp:258]     Train net output #1: loss2 = 2.02409e+07 (* 4 = 8.09637e+07 loss)
I0904 14:05:02.294389 21891 solver.cpp:571] Iteration 60, lr = 0.0001
...
...
...
I0904 14:05:53.597456 21891 solver.cpp:242] Iteration 120, loss = 3.33163e+10
I0904 14:05:53.597492 21891 solver.cpp:258]     Train net output #0: loss = 5.66763e+10 (* 1 = 5.66763e+10 loss)
I0904 14:05:53.597517 21891 solver.cpp:258]     Train net output #1: loss2 = 7.28106e+06 (* 4 = 2.91242e+07 loss)
I0904 14:05:53.597535 21891 solver.cpp:571] Iteration 120, lr = 0.0001
I0904 14:06:10.668745 21891 solver.cpp:242] Iteration 140, loss = 1.32058e+10
I0904 14:06:10.668792 21891 solver.cpp:258]     Train net output #0: loss = 5.58073e+10 (* 1 = 5.58073e+10 loss)
I0904 14:06:10.668810 21891 solver.cpp:258]     Train net output #1: loss2 = 9.37254e+06 (* 4 = 3.74902e+07 loss)
I0904 14:06:10.668829 21891 solver.cpp:571] Iteration 140, lr = 0.0001
I0904 14:06:27.734812 21891 solver.cpp:242] Iteration 160, loss = -1.63327e+09
I0904 14:06:27.734845 21891 solver.cpp:258]     Train net output #0: loss = 2.92904e+10 (* 1 = 2.92904e+10 loss)
I0904 14:06:27.734870 21891 solver.cpp:258]     Train net output #1: loss2 = 5.31872e+06 (* 4 = 2.12749e+07 loss)
I0904 14:06:27.734889 21891 solver.cpp:571] Iteration 160, lr = 0.0001
I0904 14:06:44.812893 21891 solver.cpp:242] Iteration 180, loss = -1.87728e+10
I0904 14:06:44.812940 21891 solver.cpp:258]     Train net output #0: loss = 1.63414e+10 (* 1 = 1.63414e+10 loss)
I0904 14:06:44.812959 21891 solver.cpp:258]     Train net output #1: loss2 = 2.0389e+06 (* 4 = 8.1556e+06 loss)
I0904 14:06:44.812979 21891 solver.cpp:571] Iteration 180, lr = 0.0001
```

I suspect a conversion issue in [solver.cpp](https://github.com/BVLC/caffe/blob/master/src/caffe/solver.cpp#L242) due to the large error values, because this error only occurs around 1.0e+10. However, I cannot find an obvious reason how this behavior might be introduced. The training continues and the individual loss values seem unaffected.

**Comment by ozguncck** [comment]
Could you solve your problem? I have the same problem with very small numbers. I checked the calculations of smooth_loss. Seems OK and cannot get negative mathematically but I was thinking that it is somehow related to floating point calculations and overflow.

**Comment by sunsided** [comment]
Happened to me twice today, however the loss tends to be around `-5` in my log. At one time it bounced up to `30000`, fell to `0.2` and then back to negative. I tried enabling `average_loss`, but that didn't help.

**Comment by CHUNYUWANG** [comment]
@sunsided Have you solved the problem? I met the problem.

**Comment by sunsided** [comment]
@CHUNYUWANG Sadly not. I have the feeling that this is related to too high learning rates though (maybe [gradient clipping](https://github.com/BVLC/caffe/blob/bdb94577d97da5cf5b6ec046952dbe79e9c886bf/src/caffe/proto/caffe.proto#L187) could help). The values are related to the problem at hand of course, but I could always trigger this when experimenting with base learning rates of `0.1` when `0.001` would have been enough. The overflow assumption might be correct, it seems that at some point the gradient gets to steep and this is when the combined loss goes haywire. Not sure if learning continues correctly after that, I usually stopped the process after noticing this and started over.

**Comment by CHUNYUWANG** [comment]
@sunsided Thanks for your reply. According to my understanding, loss is not directly related to learning rate. Whatever the learning rate is, the loss should not be negative. Besides, it should not be a overflow problem as a float number will not overflow to negative. I think it is related to how they compute the smoothed loss:

template <typename Dtype>

void Solver<Dtype>::UpdateSmoothedLoss(Dtype loss, int start_iter,

    int average_loss) {

  if (losses_.size() < average_loss) {

    losses_.push_back(loss);

    int size = losses_.size();

    smoothed_loss_ = (smoothed_loss_ * (size - 1) + loss) / size;

  } else {

    int idx = (iter_ - start_iter) % average_loss;

    smoothed_loss_ += (loss - losses_[idx]) / average_loss;

    losses_[idx] = loss;

  }

}


In the else part, if loss is continuously smaller than losses_[idx], smoothed_loss_ will go negative.  I don't quite understand the logic behind this function though.

**Comment by sunsided** [comment]
@CHUNYUWANG Ah yes, this is the effect I saw. When I set high learning rates, the loss would jump from some initial value in the range of `1e-1 ... 1e+2` to `1e+30` and then fall. That's the moment when it flips over.
